### ISSUE TITLE
Complete MVar2 interface

### ISSUE THREAD

**Issue by rossabaker** [typelevel/cats-effect#820]
After #739, we'll still be missing a few interesting functions present in the Haskell implementation:

* withMVar
* modifyMVar_
* modifyMVar

For consistent naming (no `MVar` suffix, avoid Scala keywords), I might consider naming these `using`, `modify_`, and `modify`, respectively.)

We should decide that we want these or reject them before releasing 2.2.0, because adding them later in the 2.x series would require the introduction of `MVar3`.

**Comment by vasilmkd** [comment]
Hello, I'd like to work on this. Should I follow the Haskell semantics with regarding exception handling? Thanks.

**Comment by alexandru** [comment]
The problem with `MVar` that we haven't foreseen is that the current `abstract class` doesn't allow adding new operations with default implementations. Adding methods with default implementations wouldn't be a problem for binary compatibility.

Instead of adding everything we ever wanted right now, maybe we should add the missing restriction:

```scala
abstract class MVar[F[_], A] {
  protected def F: Async[F]
}
```

This would allow us more flexibility in adding new operations on `MVar`.

**Comment by vasilmkd** [comment]
I like that proposal. I'll wait for a more general agreement on the whole issue then.

**Comment by rossabaker** [comment]
I think that's a good idea, but wouldn't it need to go on `MVar2` instead of `MVar`?  `MVar2` isn't locked down until 2.2.0, but `MVar` can't add anything abstract, even if it's just an instance to support more concrete methods later.

**Comment by vasilmkd** [comment]
To me, it seems that it has to go on `MVar2`, yes.

Are default methods with `(implicit F: Async[F])` parameter list out of the question?

**Comment by rossabaker** [comment]
Method-level `Async`s would be unpleasant with `MVarConcurrent`.  They should be coherent, and I suppose overrides that need `Concurrent` could ignore their method parameter and explicitly use the class constraint, but the abstract member seems a bit cleaner to me.  The cost is new operations have to go on `MVar2`.

**Comment by vasilmkd** [comment]
Thanks for clearing it up. Should I proceed with a PR?

**Comment by rossabaker** [comment]
:+1: That would be great. Thanks!

**Comment by vasilmkd** [comment]
@rossabaker @alexandru I noticed that adding a `protected def F: Async[F]` is tough when implementing `mapK` for `MVar`. I don't think it's possible to obtain an `Async[G]` from an `Async[F]` and `F ~> G`. Thoughts?

**Comment by rossabaker** [comment]
Anything that can't be implemented atop an abstract `Async[F]` can be implemented before 2.2.  I think the purpose of the abstract `F` is so we can add things for which that's sufficient without breaking compatibility.  So if a `mapK` happens before 2.2.0, I don't see a problem.

I don't have the spare brain cycles at this moment to contemplate whether the `Async[G]` can be summoned, but that's my thoughts if it can't.

**Comment by vasilmkd** [comment]
True. Unfortunately, `mapM` is already a part of the `MVar` interface, and I don't see a way to proceed with adding an abstract `Async[F]` without also providing an abstract `Async[G]` as well.

**Comment by vasilmkd** [comment]
True. Unfortunately, `mapK` is already a part of the `MVar` interface, and I don't see a way to proceed with adding an abstract `Async[F]` without also demanding an `Async[G]` in `mapK` as well, which cannot be provided as `[G[_]: Async]`, since that method doesn't exist in the interface, and thus cannot be implemented.

**Comment by rossabaker** [comment]
Let me paraphrase the problem to see if we're on the same page:

1. We add an abstract `F: Async[F]` to `MVar2`.
2. `MVarAsync` and `MVarConcurrent` are cool, because those each take an `Async[F]` already.
3. `TransformedMVar2` is not cool, because we don't have an `Async[G]`.
4. We need an `F ~> G` and `G ~> F` to derive `Async[G]` from `Async[F]`
5. All we have is `F ~> G`.
6. Go be sad on GitHub.

**Comment by vasilmkd** [comment]
@rossabaker I'd say that's a spot on summary of the situation. 😅

**Comment by rossabaker** [comment]
I'm not seeing a good way out of this.  Maybe we should push for those missing methods from Haskell's in 2.2.0, and just focus on a more extensible encoding for Cats-Effect 3?

**Comment by vasilmkd** [comment]
This might not be as simple as it seems. How can these functions be implemented for `TransformedMVar2` when their signatures are:
```scala
def use[B](f: A => G[B]): G[B] // underlying needs an `A => F[B]`

def modify[B](f: A => G[(A, B)]): G[B] // underlying needs an `A => F[(A, B)]`

def modify_[B](f: A => G[A]): G[Unit] // underlying needs an `A => F[A]`
```

**Comment by rossabaker** [comment]
Ah, right.  We're introducing invariant methods in MVar2.  Maybe that suggests an answer:

* Add `imapK` to `MVar2`.  Pass along `g` to `TransformedMVar2` constructor.

```scala
  def imapK[G[_]](f: F ~> G, g: G ~> F): MVar2[G, A]
```

* Remove the `MVar2#mapK` override, so it still returns `MVar[G, A]`, because you don't have the `g`.
* Now you can implement (I think) these new methods in `TransformedMVar2`.

It's ugly, but `MVar2` already is.  Hence the desire to not ever see an `MVar3`. :smile:

**Comment by vasilmkd** [comment]
I've been looking at the implementation of these methods [here](http://hackage.haskell.org/package/base-4.14.0.0/docs/src/Control.Concurrent.MVar.html#withMVar) and unsurprisingly, it uses masking to restore proper interruptibility of the code.

I also stumbled upon #771 and specifically this [comment](https://github.com/typelevel/cats-effect/issues/771#issuecomment-581047437).

To me, it seems that it doesn't bode well for this issue as a whole.
@rossabaker

cc @SystemFw

**Comment by SystemFw** [comment]
Yeah I wanted to let the discussion unfold, but as it stands this implementations are hard to achieve without the new interruption model. As a separate opinion, I think MVar is very error prone for users (as I've written in other issues), since it's incredible deadlock prone. Haskell has deadlock detection so you don't feel it as much, but imho in Scala is a bit of a nightmare to deal with.

**Comment by rossabaker** [comment]
I've only used it in Haskell. :laughing: 

I hadn't thought through implementing them in the untransformed types.  If they're absent for good reason, I'm happy closing this.  `MVar2` will have the similar compatibility problems to `MVar`, so I just want to be sure we can live with `MVar2` for the rest of CE2's lifespan.

**Comment by SystemFw** [comment]
> but as it stands these functions are hard to implement without the new interruption model.

I mean, they are not _impossible_ to implement, so it's really up to you. Btw I'd pretty curious if you could make a comparison between what you'd write with MVar and the approach with Ref/Deferred FSM. I'm very biased so I'd actually like to hear some pro MVar comments.

**Comment by vasilmkd** [comment]
> > but as it stands these functions are hard to implement without the new interruption model.
> 
> I mean, they are not _impossible_ to implement, so it's really up to you. Btw I'd pretty curious if you could make a comparison between what you'd write with MVar and the approach with Ref/Deferred FSM. I'm very biased so I'd actually like to hear some pro MVar comments.

I'd be willing to give this a shot. Would you be willing to review and comment on my effort?

**Comment by vasilmkd** [comment]
Any resolution for this issue or the accompanying PR?

**Comment by djspiewak** [comment]
@vasilmkd Sorry that this fell off the radar! I'll see if I can shepherd things to a resolution early this week. It's a really important thing to have a final answer on sooner rather than later, since we don't want to have `MVar2`/`MVar3`/etc in ce3 :-)

**Comment by SystemFw** [comment]
My bad! 😬 

Replied on the PR

**Comment by vasilmkd** [comment]
I believe this issue can be closed.
